src/Data/BTree/Fold.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE GADTs #-}
+module Data.BTree.Fold where
+
+import Prelude hiding (foldr, foldl)
+
+import Data.BTree.Alloc.Class
+import Data.BTree.Primitives
+
+import Control.Applicative ((<$>))
+import Control.Monad (liftM2)
+
+import Data.Monoid (Monoid, (<>), mempty)
+
+import qualified Data.Foldable as F
+import qualified Data.Map as M
+
+--------------------------------------------------------------------------------
+
+{-| Perform a right-associative fold over the tree. -}
+foldr :: (AllocM m, Key k, Value a)
+      => (a -> b -> b) -> b -> Tree k a -> m b
+foldr _ x (Tree _ Nothing) = return x
+foldr f x (Tree h (Just nid)) = foldrId (liftM2 f) (return x) h nid
+
+foldrId :: (AllocM m, Key k, Value a)
+        => (m a -> m b -> m b) -> m b -> Height h -> NodeId h k a -> m b
+foldrId f x h nid = readNode h nid >>= foldrNode f x h
+
+foldrNode :: (AllocM m, Key k, Value a)
+          => (m a -> m b -> m b) -> m b -> Height h -> Node h k a -> m b
```

Comment:
Usually you fold with a algebraic list algebra that has the type (a -> b -> m b, b). Your type is not algebraic. I don't see a real problem, but I think it's still better to go with the convention. Can you implement
foldr :: (AllocM m, Key k, Value a)
      => (a -> b -> b) -> b -> Tree k a -> m b
foldr f = foldrM (\a b -> return (f a b))

foldrM :: (AllocM m, Key k, Value a)
       => (a -> b -> m b) -> b -> Tree k a -> m b

by reusing definitions of foldrM for 'Index' and 'Map'?